src/pool.ts

```diff
@@ -0,0 +1,754 @@
+/*!
+ * This source file is part of the EdgeDB open source project.
+ *
+ * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import connect, {AwaitConnection, QueryArgs} from "./client";
+import {ConnectConfig} from "./con_utils";
+import * as errors from "./errors";
+import {LifoQueue} from "./queues";
+import {Set} from "./datatypes/set";
+
+export class Deferred<T> {
+  private _promise: Promise<T>;
+  private resolve?: (value?: T | PromiseLike<T> | undefined) => void;
+  private reject?: (reason?: any) => void;
+  private _result: T | PromiseLike<T> | undefined;
+  private _done: boolean;
+
+  public get promise(): Promise<T> {
+    return this._promise;
+  }
+
+  public get done(): boolean {
+    return this._done;
+  }
+
+  public get result(): T | PromiseLike<T> | undefined {
+    if (!this._done) {
+      throw new Error("The deferred is not resolved.");
+    }
+    return this._result;
+  }
+
+  async setResult(value?: T | PromiseLike<T> | undefined): Promise<void> {
+    while (!this.resolve) {
+      // this.resolve is configured inside a Promise callback
+      // to avoid race conditions, here we await for a next cycle until the
+      // reference is assigned
+      // Normally, awaiting a timeout of 0 in a while loop would cause high
+      // CPU usage, but in this case we expect the reference to be already
+      // populated in almost all cases, and anyway in the next cycle
+      await new Promise((resolve) => setTimeout(resolve, 0));
+    }
+    this.resolve(value);
+  }
+
+  async setFailed(reason?: any): Promise<void> {
+    while (!this.reject) {
+      // see the comment in ``setResult`` above
+      await new Promise((resolve) => setTimeout(resolve, 0));
+    }
+    this.reject(reason);
+  }
+
+  constructor() {
+    this._done = false;
+    this.reject = undefined;
+    this.resolve = undefined;
+
+    this._promise = new Promise((resolve, reject) => {
+      this.reject = reject;
+
+      this.resolve = (value?: T | PromiseLike<T> | undefined) => {
+        this._done = true;
+        this._result = value;
+        resolve(value);
+      };
+    });
+  }
+}
+
+class PoolConnectionHolder {
+  private _pool: Pool;
```

Comment:
No need for `_` for private fields. At least that's the decision I made for `client.ts` and other files, so let's keep the style consistent.

Reply:
Hm, I actually looked at my code and I see that it's not as consistent as I thought. We don't have underscores for `AwaitConnection` attributes, but then we do have them for some methods.

Given that this driver will also be used by vanilla JS code, we might want to keep the underscores here and add them to `client.ts`.

Reply:
I asked somewhere the same question in one of the other comments (whether we should keep underscores for private attributes). In some cases, having underscores make sense if you want to have, or add in the future, `public get` accessor.

I mean for example:
```
private _foo: boolean;

public get foo(): boolean { return this._foo; }
```

Reply:
Let's have underscores. Our vanilla JS users will appreciate that. And we'll add them to client.ts in a follow up PR.